provider/src/main/scala/com/actian/spark_vector/provider/Job.scala

```diff
@@ -0,0 +1,153 @@
+}
+
+/** How many streams does `Vector` expect on the node with host name `host`, at `port` */
+private[provider] case class StreamPerNode(nr: Int,
+  port: Int,
+  host: String)
+
+/**
+ * Metadata about datastreams.
+ *
+ * @param rolename The username to use when authenticating with `Vector`'s datastreams
+ * @param password The password to use when authenticating with `Vector`'s datastreams
+ * @param streams_per_node For each node that expects `Datastream` connections, this sequence will contain a `StreamPerNode` element
+ */
+private[provider] case class DataStream(rolename: String,
+  password: String,
+  streams_per_node: Seq[StreamPerNode])
+
+/**
+ * Part of a Vector external table request.
+ *
+ * @param part_id Unique identifier for this part of the request
+ * @param external_table_name The name of the external table, as defined in Vector's SQL
+ * @param operator_type The type of operation mapped to this request part. It can either be 'scan' or 'insert'.
+ * @param external_reference The reference to the external data source, e.g. file path for file sources, table names for database sources, etc.
+ * @param format What is the format of the external datasource
+ * @param column_infos Information about the columns read/inserted in the external datasource.
+ * @param options Additional options that need to be provided to the datasource
+ * @param datastream Connection (to `Vector`'s datastreams) information
+ */
+private[provider] case class JobPart(part_id: String,
+    external_table_name: String,
+    operator_type: String,
+    external_reference: String,
+    format: Option[String],
+    column_infos: Seq[ColumnInfo],
+    options: Option[Map[String, String]],
+    datastream: DataStream) {
+  def conf: VectorEndpointConf = {
+    val endpoints = for {
+      spn <- datastream.streams_per_node
+      i <- 0 until spn.nr
+    } yield VectorEndpoint(spn.host, spn.port, datastream.rolename, datastream.password)
+    VectorEndpointConf(endpoints.toIndexedSeq)
+  }
+}
+
+private[provider] object JobPart {
+  implicit val logicalTypeFormat = Json.format[LogicalType]
+  implicit val columnInfoFormat = Json.format[ColumnInfo]
+  implicit val streamPerNodeFormat = Json.format[StreamPerNode]
+  implicit val dataStreamFormat = Json.format[DataStream]
+  implicit val jobPartFormat = Json.format[JobPart]
+}
+
+/**
+ * A query request for this provider.
+ *
+ * @note Each `Job` has multiple [[parts]] that are executed sequentially
+ */
+private[provider] case class Job(transaction_id: Long, query_id: Long, `type`: String, parts: Seq[JobPart])
+
+private[provider] object Job {
+  implicit val jobFormat = Json.format[Job]
+}
+
+private[provider] sealed trait JobStatus
+
+private[provider] sealed trait JobSuccess extends JobStatus
+private[provider] case object JobSuccess extends JobSuccess {
+  implicit val successFormat = Format[JobSuccess](Reads { _.validate[JsObject] map (_ => JobSuccess) }, Writes { _ => JsObject(Nil) })
```

Comment:
DRY

maybe you could rewrite this `Format[JobSuccess](Reads { _.validate[JsObject] map (_ => JobSuccess) }, Writes { _ => JsObject(Nil) })` as a generic function within the `JobStatus` trait so that you use it here in `JobSuccess` and `JobAck` ?